Bump version number to 7 and compile old operators with old schema

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68358

Differential Revision: [D33433730](https://our.internmc.facebook.com/intern/diff/D33433730)

tugsbayasgalan committed:
Bump version number to 5 and compile old operators with old schema

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 5 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 5 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 5 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 5 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 5 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 5 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 5 and compile old operators with old schema"




[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/serialization/import_source.cpp:
```diff
@@ -578,7 +578,9 @@ void SourceImporterImpl::importClass(
       method_resolvers,
-      &self);
+      &self,
+      false,
```
Consider specifying which argument `false` is passed to like what line 578 is doing.

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1,3 +1,4 @@
+#include <caffe2/serialize/versions.h>
```
Nit:
A cpp file should first include its own header which is `python_print.h` in this case. `versions.h` should be included in the bulk section starting from line 5.

tugsbayasgalan replied:
sounds good.

gmagogsfm commented on torch/csrc/jit/serialization/python_print.cpp:
```diff
@@ -1594,7 +1588,7 @@ struct PythonPrintImpl {
 
   // The least version that supports all printed ops
-  uint64_t min_version_ = 0;
+  uint64_t min_version_ = caffe2::serialize::kProducedFileFormatVersion;
```
This is no longer `min_version`, instead it is the exact version I think. Maybe we don't even need to keep track of `min_version` as a member of this class, instead we can simply write out `caffe2::serialize::kProducedFileFormatVersion` to the `version` file.

tugsbayasgalan replied:
done

tugsbayasgalan committed:
Update on "Bump version number to 5 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan commented on torch/package/package_exporter.py:
```diff
@@ -197,7 +197,7 @@ def __init__(
 
         self.zip_file = torch._C.PyTorchFileWriter(f)
-        self.zip_file.set_min_version(6)
+        self.zip_file.set_version(6)
```
tagging @suo for this change. Since we don't rely on dynamic versioning anymore, we no longer need to take the min of different modules.

tugsbayasgalan committed:
Update on "Bump version number to 5 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan commented on caffe2/serialize/versions.h:
```diff
@@ -6,7 +6,7 @@ namespace caffe2 {
 
 constexpr uint64_t kMinSupportedFileFormatVersion = 0x1L;
-constexpr uint64_t kMaxSupportedFileFormatVersion = 0x6L;
+constexpr uint64_t kMaxSupportedFileFormatVersion = 0x7L;
```
this doesn't feel right. @cccclai @gmagogsfm Seems like mobile bytecode version relies on this? if this is not needed, I would like to remove it.

cccclai replied:
`kMaxSupportedFileFormatVersion` (global operator version) is different than `kProducedBytecodeVersion` (bytecode schema version). I don't see the bytecode schema will change with this diff, so there is no need to bump `kProducedBytecodeVersion`. @iseeyuan what do you think?

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan commented on torch/jit/operator_upgraders.py:
```diff
@@ -53,7 +53,7 @@ def div__Scalar_0_3(self: torch.Tensor, other: Union[int, float]) -> torch.Tenso
     @no_type_check
     @torch.jit.script
-    def full_names_0_4(size: List[int], fill_value: Union[int, float], *,
+    def full_0_4(size: List[int], fill_value: Union[int, float], *,
```
this is done to make sure our code works end-to-end. This shouldn't be here once we merge https://github.com/pytorch/pytorch/pull/69860

tugsbayasgalan replied:
resolved

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"




[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

cccclai commented:
Just want to double check the output version is calculated dynamically and keeps the same logic as https://github.com/pytorch/pytorch/pull/40279

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

gmagogsfm commented on test/jit/test_upgraders.py:
```diff
@@ -81,3 +92,112 @@ def f():
+        version = self._load_model_version(loaded_model)
+        self.assertTrue(version == 4)
+        loaded_model_twice = torch.jit.load(buffer)
+        # we check by its' code because graph variable names
```
s/its'/its

tugsbayasgalan replied:
done

gmagogsfm commented on test/jit/test_upgraders.py:
```diff
@@ -81,3 +92,112 @@ def f():
+        # can be different every time
+        self.assertEqual(loaded_model.code, loaded_model_twice.code)
+
+    def test_aten_test_serialization(self):
```
It is a bit hard to read and maintain such test cases. For examle, I do not know what is inside `_test_serialization_subcmul_v2.pt` to tell whether the test is correct. Could we add a TODO for later to automatically generate such test cases from structured data with ample comment for readability?

tugsbayasgalan replied:
sounds good. I added a TODO at the top of the test file marking this

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan commented on caffe2/serialize/inline_container.h:
```diff
@@ -164,7 +164,9 @@ class TORCH_API PyTorchStreamWriter final {
-  uint64_t version_ = kProducedFileFormatVersion;
+  // This number will be updated when the model has operators
+  // that have valid upgraders.
+  uint64_t version_ = caffe2::serialize::kMinProducedFileFormatVersion;
```
done

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

gmagogsfm reviewed (approved):
Looks good. Please also prepare a diff to flip the macro and let me approve it so that we can land as quickly as possible when needed.

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

gmagogsfm commented on test/cpp/jit/test_load_upgraders.cpp:
```diff
@@ -0,0 +1,45 @@
+
+  auto test_graph = loaded_m.get_method("forward").graph();
+  // should have saved with version 4, so it is still up to date
+  testing::FileCheck().check_count("aten::div", 1, true)->run(*test_graph);
```
What is this testing? Looks like the expected behavior is upgrader is not applied?

gmagogsfm commented on test/cpp/jit/test_load_upgraders.cpp:
```diff
@@ -0,0 +1,45 @@
+  auto version_map = get_operator_version_map();
+  auto upgraders = dump_upgraders_map();
+
+  for (const auto& entry : version_map) {
```
This test would pass if `version_map` is empty. Is it intended?

tugsbayasgalan replied:
Yes. This means there are no upgraders available. But I guess I can make it 1-to-1 check

gmagogsfm commented on torch/csrc/jit/operator_upgraders/upgraders_entry.cpp:
```diff
@@ -0,0 +1,73 @@
+namespace torch {
+namespace jit {
+
+static std::unordered_map<std::string, std::string> kUpgradersEntryMap(
```
Add a TODO to remove this later.

gmagogsfm commented on torch/csrc/jit/operator_upgraders/upgraders.cpp:
```diff
@@ -9,13 +12,17 @@ namespace jit {
 
 void UpgradersMap::set_content(
-    std::unordered_map<std::string, std::string>&& content) {
+    std::unordered_map<std::string, std::shared_ptr<Graph>>& content) {
```
why not pass right ref here?

tugsbayasgalan replied:
done

gmagogsfm commented on torch/csrc/jit/operator_upgraders/upgraders.cpp:
```diff
@@ -46,15 +55,16 @@ void UpgradersMap::test_only_remove_content(
 void populate_upgraders_map(
-    std::unordered_map<std::string, std::string>&& content) {
-  upgradersMap.set_content(std::move(content));
+    std::unordered_map<std::string, std::shared_ptr<Graph>> content) {
```
If we don't pass right ref+move on line 15 we can use a const ref here.

gmagogsfm commented on test/cpp/jit/test_load_upgraders.cpp:
```diff
@@ -0,0 +1,45 @@
+namespace jit {
+
+#if ENABLE_UPGRADERS
+// Basic tests to check if C++ torch::jit::load
```
Can we have a test case that checks C++ API can load and upgrade an old model?

gmagogsfm commented:
Awesome, really quick turn-around. Overall looks good, just a few comments.

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33198155](https://our.internmc.facebook.com/intern/diff/D33198155)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33198155).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"



[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"



[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33433730).

tugsbayasgalan committed:
Update on "Bump version number to 7 and compile old operators with old schema"


Differential Revision: [D33433730](https://our.internmc.facebook.com/intern/diff/D33433730)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33433730).
